### ISSUE TITLE
ModelBuilder API: Please don't remove database-agnostic API

### ISSUE THREAD

**Issue by joshmouch** [dotnet/efcore#3362]
So I'm in the process of migration a large EF6 code first database project to an EF7 one and I have to admit that I'm a bit taken aback.  One of the primary goals of code first was to be database-agnostic.  Theoretically, you could simply change the database server in use and it would "just work".  That is no longer true.  When switching to EF7, for about 90% of my entity properties, I now need to give code first a SQL Server specific type to use (e.g. "varchar(200)").

My most recent complaint is that PropertyBuilder<string>.IsUnicode() is gone.  I read the design notes here:

https://github.com/aspnet/EntityFramework/wiki/Entity-Framework-Design-Meeting-Notes-June-5,-2014#decision-property-facet-apis

and I don't agree at all.

From the document:

> "The idea of something being Unicode or not isn't really a generic concept."

This actually _is_ a generic concept or at least is close to a generic concept.  .NET always stores strings as UTF16, however, long-term storage rarely ever does.  It uses ANSI, UTF8, UTF16, UTF32, etc.  You need a way to represent your desire to store the string in a more compact way than .NET does.  If the database doesn't support it then who cares?  ...just fall back to UTF16.

I've been around the block enough times to realize that the true motivation behind dropping so many of the EF6 features is to get EF7 out faster - which I can understand.  But please try to keep at least some of the database-agnostic features.

In the short term, it may seem like appending .HasColumnType("varchar(200)") after every property is convenient, but what happens when Microsoft comes out with Sql Server 2020 Flash Drive edition, which no longer supports UTF8 data types or even variable column lengths?   You suddenly have a TON of .NET users who can't use that database without significant code change.  You're just lucky that all of the Microsoft SQL Server editions are so similar in their data types.

**Comment by joshmouch** [comment]
Sorry that the previous comment is a bit wordy.  I just want to make sure that you keep the database-agnostic goal of code-first.

**Comment by rowanmiller** [comment]
**Triage:** Discussed and we agree. Opened https://github.com/aspnet/EntityFramework/issues/3420 to track adding `IsUnicode`. Was there any other provider agnostic APIs that you found were missing?

**Comment by joshmouch** [comment]
I think IsFixedLength() is probably the other one that is agnostic yet can have an easy fallback if not supported by the provider.  I personally don't use it, but I do use Microsoft Identity 2.0, and it uses that quite a bit.
